esp-hal/src/i2c.rs

```diff
@@ -758,14 +821,21 @@ mod asynch {
         async fn wait_for_completion(&self, end_only: bool) -> Result<(), Error> {
             self.peripheral.check_errors()?;
 
+            // ideally we should have a timeout here like we have for the blocking case
```

Comment:
For ESP32 only (apparently) there is the same problem which is fixed for blocking via MAX_ITERATIONS but it's a bit of a problem for async. Ideally, I don't want to waste a TIMG timer or SYSTIMER alarm for this. I also think other drivers could benefit from something like this.

Maybe CPU internal timers given the timeouts are very short? In any case I think we should do that improvement in another PR since the changes here already help a lot

Reply:
A small amount of blocking is fine IMO, definitely better than getting stuck. Using a TIMG or SYSTIMER alarm will probably be even slower given the async context switch involved.

Reply:
I came up with a special solution for ESP32 - not proud of it but it works and doesn't waste a timer